### PR TITLE
Allow expandable fields to be writable

### PR DESCRIPTION
**Pull request by gareth-lloyd**
## Todo

- [x] Check edge cases
- [x] Tests - keep coverage at 100
- [x] Documentation

### CODE REVIEW

File: rest_framework_serializer_extensions/fields.py [hunk 1]
```diff
@@ -59,15 +50,24 @@ class HashIdField(GetHashIdModelMixin, Field):
...
         )
 
     def to_internal_value(self, value):
-        return utils.internal_id_from_model_and_external_id(
-            self.get_model(), value
-        )
+        model = self.get_model()
+        try:
+            return utils.internal_id_from_model_and_external_id(model, value)
+        except ContentType.DoesNotExist:
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/fields.py]: I think you could simply except `django.core.exceptions.ObjectDoesNotExist` and return malformed in both cases?

File: rest_framework_serializer_extensions/serializers.py [hunk 2]
```diff
@@ -409,6 +429,21 @@ def _get_expand_fields(self, standard_fields):
...
+        """
+        validated_data = (
+            super(ExpandableFieldsMixin, self).run_validation(data=data)
+        )
+        for id_field_name in self._id_fields_to_translate:
+            if id_field_name in validated_data:
+                model_field_name = id_field_name[:-3]
+                validated_data[model_field_name] = (
+                    validated_data.pop(id_field_name)
+                )
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/serializers.py]: It looks like you'll get a problem here if you *also* expand the field normally during expansion. If the view set an `extensions_expand = {'x'}` attribute, then x will have a collision.

I think it's probably worth thinking over the situations where you might want to post expanded fields. Granted, it seems weird that you would want to update an attribute on a foreign relation from a single post, but as you don't *only* expand foreign relations, there may be legitimate use cases for doing so.

File: rest_framework_serializer_extensions/utils.py [hunk 3]
```diff
@@ -71,3 +73,31 @@ def internal_id_from_model_and_external_id(model, external_id):
...
+        model = model_definition
+
+    try:
+        assert issubclass(model, models.Model)
+    except (AssertionError, TypeError):
+        raise AssertionError(
+            '"{0}"" is not a Django model'.format(model_definition)
+        )
+
+    return model
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/utils.py]: Looks good to me. I think a few tests cases showing how this work and how extensions deals with deserialization in general would be useful in highlighting any use cases where this would be problematic

File: docs/usage-serializers.md [hunk 4]
```diff
@@ -137,6 +137,49 @@ class OwnerSerializer(SerializerExtensionsMixin, ModelSerializer):
...
+                read_only=False
+            )
+        )
+```
+
+As usual, the above serializer will have an `organization_id` field by default.
+To set or alter the organization of an owner, pass a value to this field:
+
+```py
+serializer = OwnerSerializer(data=dict(name="Elliot", organization_id=2))
```
Comment (evenicoulddoit) [inline docs/usage-serializers.md]: 👍 , this might be clearer (and more consistent with other documentation) by using this:

```js
>>> POST /owner/2/
{
  "name": "Elliot",
  "organization_id": 2
}
```

File: docs/usage-serializers.md [hunk 5]
```diff
@@ -137,6 +137,49 @@ class OwnerSerializer(SerializerExtensionsMixin, ModelSerializer):
...
+
+During validation the serializer will find the corresponding organization and
+add make it available in the deserialized data under the key
+`organization_id_resolved`:
+
+```py
+{
+    "name": "Elliot",
+    "organization_id": 2,
+    "organization_id_resolved": <Organization: Allsafe>
```
Comment (evenicoulddoit) [inline docs/usage-serializers.md]: 👍 for continuing the dummy data theme :P

File: docs/usage-serializers.md [hunk 6]
```diff
@@ -137,6 +137,49 @@ class OwnerSerializer(SerializerExtensionsMixin, ModelSerializer):
 Setting `id_source=False` results in no ID field being included.
 
 
+### Writable ForeignKey relationships
```
Comment (evenicoulddoit) [inline docs/usage-serializers.md]: As below, maybe this para should have a ref to (i.e. for POST/PUT requests)

File: rest_framework_serializer_extensions/serializers.py [hunk 7]
```diff
@@ -288,7 +292,8 @@ def get_expand_id_field(self, field_name, field_definition):
         if hasattr(self, 'get_{0}_id'.format(field_name)):
             return serializers.SerializerMethodField(source='*')
 
-        kwargs = dict(read_only=True)
+        read_only = field_definition.get('read_only', True)
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/serializers.py]: I think it would be nice to put this in `_standardise_expandable_definition()` so that we can have some assurance that `read_only` exists. Ultimately I think we should make an `ExpandableField` class to make the interface slightly nicer.

File: rest_framework_serializer_extensions/serializers.py [hunk 8]
```diff
@@ -377,16 +390,23 @@ def _get_expand_fields(self, standard_fields):
...
             if (
                 not field_definition.get('many') and
                 field_definition.get('id_source') is not False
             ):
-                expanded_fields['{0}_id'.format(field_name)] = (
+                id_field_name = '{0}_id'.format(field_name)
+                expanded_fields[id_field_name] = (
                     self.get_expand_id_field(field_name, field_definition)
                 )
+                if not field_definition.get('read_only'):
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/serializers.py]: As above, you can then just do `if not field_definition['read_only']`

File: rest_framework_serializer_extensions/serializers.py [hunk 9]
```diff
@@ -409,6 +429,23 @@ def _get_expand_fields(self, standard_fields):
...
+        validated_data = (
+            super(ExpandableFieldsMixin, self).run_validation(data=data)
+        )
+        for id_field_name in self._id_fields_to_translate:
+            if id_field_name in validated_data:
+                # ID field has been resolved to an instance
+                instance = validated_data[id_field_name]
+                resolved_field_name = '{0}_resolved'.format(id_field_name)
+                validated_data[resolved_field_name] = instance
+                validated_data[id_field_name] = instance.pk
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/serializers.py]: Might be worth commenting that you're translating the ID back to an ID?

File: rest_framework_serializer_extensions/fields.py [hunk 10]
```diff
@@ -59,15 +50,21 @@ class HashIdField(GetHashIdModelMixin, Field):
...
             self.get_model(), value
         )
 
     def to_internal_value(self, value):
-        return utils.internal_id_from_model_and_external_id(
-            self.get_model(), value
-        )
+        model = self.get_model()
+        try:
+            return utils.internal_id_from_model_and_external_id(model, value)
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/fields.py]: This isn't *necessarily* true from what I can see. I.e. it could be a perfectly valid HashID, but just not correspond to the model we're expecting?

File: tests/test_utils.py [hunk 11]
```diff
@@ -151,3 +151,32 @@ def test_returns_valid_internal_id(self):
...
+            utils.model_from_definition('tests.models.CarModel')
+        )
+
+    def test_pass_model_class_as_model_definition(self):
+        """
+        Test that the method returns its argument when Django model is passed.
+        """
+        self.assertEquals(
+            models.CarModel, utils.model_from_definition(models.CarModel)
+        )
```
Comment (evenicoulddoit) [inline tests/test_utils.py]: 👍 Awesomesauce. Looks great, just a few nitpicks

File: rest_framework_serializer_extensions/serializers.py [hunk 12]
```diff
@@ -440,8 +446,12 @@ def run_validation(self, data=empty):
             if id_field_name in validated_data:
                 # ID field has been resolved to an instance
                 instance = validated_data[id_field_name]
+
+                # Create a new field to contain the resolved instance.
                 resolved_field_name = '{0}_resolved'.format(id_field_name)
                 validated_data[resolved_field_name] = instance
+
+                # Translate ID field contents back to an ID
                 validated_data[id_field_name] = instance.pk
```
Comment (evenicoulddoit) [inline rest_framework_serializer_extensions/serializers.py]: Just realised this will always be in the internal ID, but I think that's fine
